### PR TITLE
Remove unwraps in StateMachine

### PR DESCRIPTION
**Pull request by usbalbin**
This PR is focused on #908 and replaces some unwraps from StateMachine with some error propogation.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -213,7 +213,8 @@ impl<'a, T, E: Send + Sync + Clone + 'static> Application<'a, T, E> {
         #[cfg(feature = "profiler")]
         profile_scope!("initialize");
         self.states
-            .start(StateData::new(&mut self.world, &mut self.data));
+            .start(StateData::new(&mut self.world, &mut self.data))
+            .unwrap(); //TODO: figure out whether this needs to be handled
```
Comment (usbalbin) [inline src/app.rs]: `StateMachine::start` errors when there are no states available. If I understand the situation this only happens when it is run after it has previously been stopped by `stop` or `pop` until empty. Do we care about those situations here? Will a game ever be run multiple times?

Comment (LucioFranco) [reply src/app.rs]: I would wrap this with a message that notifies the user that they initialized without states, I would use `expect_err` in this case.

Comment (usbalbin) [reply src/app.rs]: Wouldn't that be just plain `expect(“some message”)`? Or did I misunderstand?
```rust
self.states
    .start(start(StateData::new(&mut self.world, &mut self.data))
    .expect_err(“Tried to start state machine without any states present”);
```
Panics every time except when `start` returns an `Err`

Comment (LucioFranco) [reply src/app.rs]: Yup, you are right, somehow slept read that doc 😄 . Yeah, so it should be just `expect`.

File: src/state.rs [hunk 2]
```diff
@@ -2,8 +2,26 @@
 
 use amethyst_input::is_close_requested;
 use ecs::prelude::World;
+use std::fmt::Result as FmtResult;
+use std::fmt::{Display, Formatter};
 use {GameData, StateEvent};
 
+#[derive(Debug)]
+pub enum StateError {
```
Comment (jojolepro) [inline src/state.rs]: Missing documentation explaining what it is. (its obvious, but we prefer to have docs as much as possible)

File: src/state.rs [hunk 3]
```diff
@@ -272,12 +290,16 @@ impl<'a, T, E: Send + Sync + 'static> StateMachine<'a, T, E> {
     ///
     /// # Panics
```
Comment (jojolepro) [inline src/state.rs]: You forgot to update the documentation. It doesn't panic anymore but returns the StateError

**Comment by usbalbin** [comment]
I hope I put my entry under the right section in the changelog

File: src/state.rs [hunk 4]
```diff
@@ -342,6 +362,8 @@ impl<'a, T, E: Send + Sync + 'static> StateMachine<'a, T, E> {
             }
 
             self.state_stack.push(state);
+
+            //State was just pushed, thus pop will always succeed
```
[the remark this answers is not in the document]
Comment (usbalbin) [reply src/state.rs]: Oh okay, I'll try to remember to next time:)

**Comment by LucioFranco** [comment]
This looks good to go! Thank you for the contribution, look forward to seeing some more 😉 😄 .

git merge

bors r+

**Pull request merged**